query/src/interpreters/async_insert_queue.rs

```diff
@@ -0,0 +1,419 @@
+}
+
+impl AsyncInsertQueue {
+    pub fn try_create(
+        // TODO(fkuner): maybe circular reference
+        session_mgr: Arc<RwLock<Option<Arc<SessionManager>>>>,
+        runtime: Arc<Runtime>,
+        max_data_size: u64,
+        busy_timeout: Duration,
+        stale_timeout: Duration,
+    ) -> Self {
+        Self {
+            session_mgr,
+            runtime,
+            max_data_size,
+            busy_timeout,
+            stale_timeout,
+            queue: Arc::new(RwLock::new(Queue::default())),
+            current_processing_insert: Arc::new(RwLock::new(QueryIdToEntry::default())),
+        }
+    }
+
+    pub async fn start(self: Arc<Self>) {
+        let self_arc = self.clone();
+        // busy timeout
+        let busy_timeout = self_arc.busy_timeout;
+        self_arc.clone().runtime.as_ref().inner().spawn(async move {
+            let mut intv = interval(busy_timeout);
+            loop {
+                intv.tick().await;
+                let timeout = self_arc.clone().busy_check();
+                intv = interval(timeout);
+            }
+        });
+        // stale timeout
+        let stale_timeout = self.stale_timeout;
+        if !stale_timeout.is_zero() {
+            self.clone().runtime.as_ref().inner().spawn(async move {
+                let mut intv = interval(stale_timeout);
+                loop {
+                    intv.tick().await;
+                    self.clone().stale_check();
+                }
+            });
+        }
+    }
+
+    pub async fn push(
+        self: Arc<Self>,
+        plan_node: Arc<InsertPlan>,
+        ctx: Arc<QueryContext>,
+    ) -> Result<()> {
+        let self_arc = self.clone();
+        let plan = plan_node.clone();
+        let settings = ctx.get_changed_settings();
+
+        let data_block = match &plan_node.source {
+            common_planners::InsertInputSource::SelectPlan(plan) => {
+                let select_interpreter = SelectInterpreter::try_create(ctx.clone(), SelectPlan {
+                    input: Arc::new((**plan).clone()),
+                })
+                .unwrap();
+
+                let mut pipeline = select_interpreter.create_new_pipeline().unwrap();
+
+                let mut sink_pipeline_builder = SinkPipeBuilder::create();
+                for _ in 0..pipeline.output_len() {
+                    let input_port = InputPort::create();
+                    sink_pipeline_builder.add_sink(
+                        input_port.clone(),
+                        MemoryTableSink::create(input_port, ctx.clone()),
+                    );
+                }
+                pipeline.add_pipe(sink_pipeline_builder.finalize());
+                let executor =
+                    PipelineCompleteExecutor::try_create(self.runtime.clone(), pipeline).unwrap();
+                executor.execute().unwrap();
+                drop(executor);
+                let blocks = ctx.consume_precommit_blocks();
```

Comment:
This is incorrect, the `precommit_blocks` may not have the same schema as insert.

Reply:
Take a look at FuseSink.